UniversalAlgebra/Structures/AsRecordsHoms.lagda

```diff
@@ -0,0 +1,223 @@
+  →            is-hom-op 𝑨 𝑩 f → is-hom-op 𝑩 𝑪 g → is-hom-op 𝑨 𝑪 (g ∘ f)
+ ∘-is-hom-op {f}{g} fho gho 𝑓 a = cong g (fho 𝑓 a) ∙ gho 𝑓 (f ∘ a)
+
+ ∘-is-hom : {f : (carrier 𝑨) → (carrier 𝑩)}{g : (carrier 𝑩) → (carrier 𝑪)}
+  →         is-hom 𝑨 𝑩 f → is-hom 𝑩 𝑪 g → is-hom 𝑨 𝑪 (g ∘ f)
+ ∘-is-hom {f} {g} fhro ghro = ihr , iho
+  where
+  ihr : is-hom-rel 𝑨 𝑪 (g ∘ f)
+  ihr = ∘-is-hom-rel {f}{g} (fst fhro) (fst ghro)
+
+  iho : is-hom-op 𝑨 𝑪 (g ∘ f)
+  iho = ∘-is-hom-op {f}{g} (snd fhro) (snd ghro)
+
+ ∘-hom : hom 𝑨 𝑩  →  hom 𝑩 𝑪  →  hom 𝑨 𝑪
+ ∘-hom (f , fh) (g , gh) = g ∘ f , ∘-is-hom {f}{g} fh gh
+
+
+module _ {𝑅 𝐹 : signature} where -- (𝑨 : structure {α} 𝑅 {ρ₀} 𝐹){𝑩 : structure {β} 𝑅 {ρ₁} 𝐹}(𝑪 : structure {γ} 𝑅 {ρ₂} 𝐹) where
+ 𝒾𝒹 : (𝑨 : structure {α} {ρ} 𝑅 𝐹) → hom 𝑨 𝑨
+ 𝒾𝒹 _ = id , (λ R a z → z)  , (λ f a → refl)  -- (λ R a → refl)
+
+module _ {𝑅 𝐹 : signature} (𝑨 : structure {α} {ρ₀} 𝑅 𝐹)(𝑩 : structure {β} {ρ₁} 𝑅 𝐹) where
+
+ is-mon : ((carrier 𝑨) → (carrier 𝑩)) → Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ is-mon g = is-hom 𝑨 𝑩 g × IsInjective g
+
+ mon : Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ mon = Σ[ g ∈ ((carrier 𝑨) → (carrier 𝑩)) ] is-mon g
+
+ mon→hom : mon → hom 𝑨 𝑩
+ mon→hom ϕ = (fst ϕ) , fst (snd ϕ )
+
+
+ is-epi : ((carrier 𝑨) → (carrier 𝑩)) → Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ is-epi g = is-hom 𝑨 𝑩 g × IsSurjective g
+
+ epi : Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ epi = Σ[ g ∈ ((carrier 𝑨) → (carrier 𝑩)) ] is-epi g
+
+ epi→hom : epi → hom 𝑨 𝑩
+ epi→hom ϕ = (fst ϕ) , fst (snd ϕ)
+
+module _ {𝑅 𝐹 : signature} where
+ open Lift
+
+ 𝓁𝒾𝒻𝓉 : {𝑨 : structure {α} {ρ} 𝑅 𝐹} → hom 𝑨 (Lift-structure 𝑨 β)
+ 𝓁𝒾𝒻𝓉 = lift , (λ R a x → x) , λ f a → refl
+
+ 𝓁ℴ𝓌ℯ𝓇 : {α β : Level}{𝑨 : structure {α} {ρ} 𝑅 𝐹} → hom (Lift-structure 𝑨 β) 𝑨
+ 𝓁ℴ𝓌ℯ𝓇 = lower , (λ R a x → x) , (λ f a → refl)
+
+-- Kernels of homomorphisms
+
+
+open ≡-Reasoning
+module _ {𝑅 𝐹 : signature} {wd : swelldef ℓ₀ β}{𝑨 : structure {α} {β ⊔ ρ₀} 𝑅 𝐹}{𝑩 : structure {β} {ρ₁} 𝑅 𝐹} where
+
+ homker-comp : (h : hom 𝑨 𝑩) → compatible 𝑨 (ker (fst h))
+ homker-comp h f {u}{v} kuv = ((fst h) (((op 𝑨)f) u))  ≡⟨(snd (snd h)) f u ⟩
+                              ((op 𝑩) f)((fst h) ∘ u) ≡⟨ wd ((op 𝑩)f) ((fst h) ∘ u) ((fst h) ∘ v) kuv ⟩
+                              ((op 𝑩) f)((fst h) ∘ v) ≡⟨((snd (snd h)) f v)⁻¹ ⟩
+                              (fst h)(((op 𝑨)f) v)   ∎
+
+ kerlift-comp : (h : hom 𝑨 𝑩) → compatible 𝑨 (kerlift (fst h) (α ⊔ ρ₀) )
+ kerlift-comp (h , hhom) f {u}{v} kuv = lift goal
+  where
+  goal : h (op 𝑨 f u) ≡ h (op 𝑨 f v)
+  goal = h (op 𝑨 f u)    ≡⟨ snd hhom f u ⟩
+         (op 𝑩 f)(h ∘ u) ≡⟨ wd (op 𝑩 f)(h ∘ u)(h ∘ v)(lower ∘ kuv) ⟩
+         (op 𝑩 f)(h ∘ v) ≡⟨ (snd hhom f v)⁻¹ ⟩
+         h (op 𝑨 f v)    ∎
+
+ open import Structures.AsRecordsCongruences --  {𝑅 = 𝑅}{𝐹 = 𝐹}
+
+ kercon : hom 𝑨 𝑩 → con 𝑨
+ kercon (h , hhom) = ((λ x y → Lift (α ⊔ ρ₀) (h x ≡ h y)) , goal) , kerlift-comp (h , hhom)
+  where
+  goal : IsEquivalence (λ x y → Lift (α ⊔ ρ₀) (h x ≡ h y))
+  goal = record { refl = lift refl
```

Comment:
All this explicit lifting strikes me as a design smell. My experience is that needing to lift is a symptom that the definitions are not quite right.  I know this is a vague comment, but I've definitely changed definitions in `agda-categories` to avoid lifting -- then showed how the traditional definition corresponded to the new one plus some pesky liftings.  Adjoints come to mind as one particular such example.

Reply:
I agree... but somehow all the stuff around kernels of homs, which worked out okay for Algebras, is harder with structures.  I'll keep working on it and hopefully be able to improve and do away with lifting.

Reply:
I'm guessing that going to Setoid will help, as the various things related to quotients will simplify, and add a parameter where currently multiple levels end up 'colliding'.